### ISSUE TITLE
Rule proposal: Disallow unused JSX expressions

### ISSUE THREAD

**Issue by Daniel15** [yannickcr/eslint-plugin-react#510]
There's probably a better name for this, I just couldn't think of it :stuck_out_tongue: 

These are valid:

``` js
function foo() {
  return (
    <div>Hello world!</div>
  );
}
```

``` js
var x = <div>Hello world!</div>;
```

These are invalid:

``` js
function foo() {
  <div>Hello world!</div>;
}
```

``` js
if (something) {
  <div>Hello world!</div>;
}
```

While doing a code review, I encountered some code that missed a `return` statement and just had the JSX sitting there. No lint error :(

**Comment by ljharb** [comment]
+1, this should parallel http://eslint.org/docs/rules/no-unused-expressions

**Comment by Daniel15** [comment]
Yeah, that name is much nicer :+1:

**Comment by Daniel15** [comment]
Looks like this was implemented as `require-render-return`, at least for the simple cases.

**Comment by ljharb** [comment]
I still think it would be useful on its own.

**Comment by Daniel15** [comment]
I hit something similar again today:
```javascript
      {sections.map(section => {
        <AdsHelpTrayContentList
          contentList={section.items}
          key={section.title}
          title={section.title}
          onItemSelected={props.onItemSelected}
        />;
      })}
```

Can you see the problem?

...

I used `{ }` instead of `( )` in the arrow function for the `map` call. 😠

**Comment by golopot** [comment]
Since eslint core rule `no-unused-expressions` has covered jsx expressions now, is this proposal still relevant?

**Comment by ljharb** [comment]
@golopot when was that added?

**Comment by golopot** [comment]
@ljharb [demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogMiAqL1xuXG5jb25zdCBmb28gPSAoKSA9PiB7IDxkaXYgLz4gfSAiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19) .

**Comment by ljharb** [comment]
I see that it works, but since https://github.com/eslint/eslint/pull/2030 eslint is explicitly supposed to not care about variable usage inside jsx, so I’m wondering when it started checking unused jsx expressions.

**Comment by golopot** [comment]
I think that the commit you referred to only affects `no-unused`.  The implementation of `no-unused-expressions` does not rely on `.eslintUsed` property.

**Comment by ljharb** [comment]
ok, so the implication is that this capability has always existed in eslint and this rule wasn’t needed.

Closing, but will reopen if @Daniel15 confirms there’s still a gap.

**Comment by Daniel15** [comment]
Thanks! I'll look into `no-unused-expressions`. Look like at Facebook we had originally disabled that rule because it showed a warning for `foo && foo.bar()`, but it looks like the `allowShortCircuit` config option can be set to allow that. I'll try it out.
